db/syncPostsToGrapher.ts

```diff
@@ -99,6 +104,78 @@ export async function buildReusableBlocksResolver(): Promise<BlockResolveFunctio
         replaceReusableBlocksRecursive(content, replacerFunction)
 }
 
+export const postLinkCompareStringGenerator = (item: PostLink): string =>
+    `${item.linkType} - ${item.target} - ${item.hash} - ${item.queryString}`
+
+export function getLinksToAddAndRemoveForPost(
+    post: PostRow,
+    existingLinksForPost: PostLink[],
+    content: string,
+    postId: number
+): { linksToAdd: PostLink[]; linksToDelete: PostLink[] } {
+    const linksInDb = groupBy(
+        existingLinksForPost,
+        postLinkCompareStringGenerator
+    )
+
+    const allHrefs = excludeNullish(
+        [...content.matchAll(anyHrefRegex)].map((x) =>
+            x.groups?.["url"]
+                ? {
+                      url: x.groups?.["url"].substring(0, 2046),
+                      sourceId: postId,
+                      componentType: "href",
+                  }
+                : undefined
+        )
+    )
+    const allSrcs = excludeNullish(
+        [...content.matchAll(anySrcRegex)].map((x) =>
+            x.groups?.["url"]
+                ? {
+                      url: x.groups?.["url"].substring(0, 2046),
+                      sourceId: postId,
+                      componentType: "src",
+                  }
+                : undefined
+        )
+    )
+    const allProminentLinks = excludeNullish(
+        [...content.matchAll(prominentLinkRegex)].map((x) =>
+            x.groups?.["url"]
+                ? {
+                      url: x.groups?.["url"].substring(0, 2046),
+                      sourceId: postId,
+                      componentType: "prominent-link",
+                  }
+                : undefined
+        )
+    )
+    const linksInDocument = groupBy(
+        [
+            ...allHrefs.map((link) => PostLink.createFromUrl(link)),
+            ...allSrcs.map((link) => PostLink.createFromUrl(link)),
+            ...allProminentLinks.map((link) => PostLink.createFromUrl(link)),
+        ],
+        postLinkCompareStringGenerator
+    )
+
+    const linksToAdd: PostLink[] = []
+    const linksToDelete: PostLink[] = []
+
+    // This is doing a set difference, but we want to do the set operation on a subset
```

Comment:
There's the slight technicality in here that the same link can appear multiple times in the same document, which is something we don't detect here.
This is fine for our current use case. And is probably also fine overall (especially seeing that WP is gonna be dead soon anyways).

If you agree, we can also change out the `groupBy` calls to `keyBy`.